python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+import heapq
+import random
+from typing import Any, Callable, Iterator, List, Optional, Tuple
+
+from ray.types import ObjectRef
+from ray.data.block import Block, BlockMetadata, BlockAccessor, DataBatch
+from ray.data._internal.block_batching.interfaces import (
+    Batch,
+    LogicalBatch,
+    BlockPrefetcher,
+)
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.memory_tracing import trace_deallocation
+
+
+def bundle_block_refs_to_logical_batches(
+    block_ref_iterator: Iterator[Tuple[ObjectRef[Block], BlockMetadata]],
+    batch_size: Optional[int],
+    drop_last: bool = False,
+) -> Iterator[LogicalBatch]:
+    """Given an iterator of block object references, and their corresponding metadata,
+    bundles the block object references into groups of the provided `batch_size`.
+
+    The output iterator returns an iterator over LogicalBatch objects.
+
+    This function does not do any slicing or creation of actual batch objects.
+    """
+    batch_buffer: List[ObjectRef[Block]] = []
+    buffer_size = 0
+    starting_index = 0
+
+    global_index = 0
+    num_rows_in_last_block = 0
+
+    if batch_size is None:
+        for block_ref, metadata in block_ref_iterator:
+            yield LogicalBatch(global_index, [block_ref], 0, None, metadata.num_rows)
+            global_index += 1
+    else:
+        while True:
+            if buffer_size < batch_size:
+                # Pull next block from iterator if current buffer is not enough to fill
+                # a batch.
+                try:
+                    block_ref, metadata = next(block_ref_iterator)
+                except StopIteration:
+                    break
+                batch_buffer.append(block_ref)
+                buffer_size += metadata.num_rows
+                num_rows_in_last_block = metadata.num_rows
+
+            if buffer_size == batch_size:
+                # If equal to batch size, then yield the full buffer.
+                yield LogicalBatch(
+                    global_index, batch_buffer, starting_index, None, buffer_size
+                )
+                batch_buffer = []
+                buffer_size = 0
+                starting_index = 0
+                num_rows_in_last_block = 0
+                global_index += 1
+
+            if buffer_size > batch_size:
```

Comment:
Should this be an elif? Otherwise it's slightly confusing since you do mutate buffer_size in the block above.

Reply:
This is no longer needed with the new code changes